query/query.go

```diff
@@ -844,6 +914,115 @@ func (q *ParsedGroupedQuery) computeNewCalculationResultArrayFn() calculation.Ne
 	}
 }
 
+// ParsedTimeBucketQuery represents a validated, sanitized time bucket query.
+// Query results are a list of time buckets sorted by time in ascending order,
+// each which contains the count of documents whose timestamps fall in the bucket.
+type ParsedTimeBucketQuery struct {
+	Namespace           string
+	StartNanosInclusive int64
+	EndNanosExclusive   int64
+	Granularity         time.Duration
+	Filters             []FilterList
+
+	// Derived fields.
+	FieldConstraints map[hash.Hash]FieldMeta // Field constraints inferred from query
+}
+
+func newParsedTimeBucketQuery(q *ParsedQuery) (ParsedTimeBucketQuery, error) {
+	tbq := ParsedTimeBucketQuery{
+		Namespace:           q.Namespace,
+		StartNanosInclusive: q.StartTimeNanos,
+		EndNanosExclusive:   q.EndTimeNanos,
+		Granularity:         *q.TimeGranularity,
+		Filters:             q.Filters,
+	}
+	if err := tbq.computeDerived(q.opts); err != nil {
+		return ParsedTimeBucketQuery{}, err
+	}
+	return tbq, nil
+}
+
+// NewTimeBucketResults creats a new time bucket results from the parsed time bucket query.
+func (q *ParsedTimeBucketQuery) NewTimeBucketResults() *TimeBucketResults {
+	bucketSizeNanos := q.Granularity.Nanoseconds()
+	startInclusiveBucketNanos := q.StartNanosInclusive / bucketSizeNanos * bucketSizeNanos
```

Comment:
doesn't 1/bucketSizeNanos * bucketSizeNanos cancel out?

Reply:
oh, you're removing the trailing nanos... a comment would be nice

Reply:
Sure can add comments